File: LatinSq_Whitening.m
function output = LatinSq_Whitening(input,L,opt)
%=============================================================
% FUNCTION: LatinSq_Whitening
% -- Generates n Latin square of order d dependent on key K
% Input:
%       input = a 256x256 matrix, 
%           L = a 256x256 Latin square with symbol set {0,1,...,255}
% Ouptut:
%      output = a 256x256 matrix
%=============================================================
% BSD licence 
%=============================================================
% By Yue (Rex) Wu (ywu03@tufts.ece.edu)
% ECE Department, Tufts University
% Mar 10, 2012
%=============================================================

if ~exist('opt','var')
    opt = 'encryption';
end

switch opt
    case 'encryption'
        switch mod(L(1),3)
            case 1
                input = flipud(input);
            case 2
                input = fliplr(input);
        end
        output = mod(input+L,256);
    case 'decryption'
        output = mod(input-L,256);
        switch mod(L(1),3)
            case 1
                output = flipud(output);
            case 2
                output = fliplr(output);
        end
end
        
        

File: LatinSq_Whitening.py
#=============================================================
# FUNCTION: LatinSq_Whitening
# -- Generates n Latin square of order d dependent on key K
# Input:
#      inputMatrix = a 256x256 matrix, 
#                L = a 256x256 Latin square with symbol set {0,1,...,255}
#                opt = encryption/decryption
# Output:
#      output = a 256x256 matrix
#=============================================================

from numpy import *

def LatinSq_Whitening(inputMatrix,L,opt):
    if opt == None: opt = 'encryption'

    if opt == 'encryption':
        if L[0]%3 == 1: inputMatrix = flipud(inputMatrix)
        if L[0]%3 == 2: inputMatrix = fliplr(inputMatrix)
        output = (inputMatrix+L)%256

    if opt == 'decryption':
        output = (inputMatrix-L)%256
        if L[0]%3 == 1: output = flipud(output)
        if L[0]%3 == 2: output = fliplr(output)
    return output